internal/common/tree.go

```diff
@@ -176,84 +239,59 @@ func RemoveLabelFromTree(node *DNSTreeNode, label string) {
 }
 
 func AddLabelToTree(node *DNSTreeNode, label, value string) {
+	if len(node.Endpoints) == 0 {
+		node.Endpoints = append(node.Endpoints, &endpoint.Endpoint{Labels: map[string]string{label: value}})
+	} else {
+		for _, ep := range node.Endpoints {
+			if ep.Labels == nil {
+				ep.Labels = endpoint.NewLabels()
+			}
+			ep.Labels[label] = value
+		}
+	}
 	for _, c := range node.Children {
 		AddLabelToBranch(node, c.Name, label, value)
```

Comment:
The funny thing is that with this line commented out, all tests but one succeed. The `AddLabelToBranch` is basically "label the node" and `AddLabelToTree` is "label everything under including the node". I don't think that matters, tbh but if we migrate the "create new endpoint" bit to the `AddLabelToTree` this should do the same job with half as much code :thinking:

Reply:
I've played around with what you're suggesting, but I can't seem to find a way to make that work. Let's leave it for now I think, as we are running out of time.